peps/pep-0793.rst

```diff
@@ -0,0 +1,476 @@
+
+A new ``PyType_GetModuleByToken`` function will be added, with a signature
+like ``PyType_GetModuleByDef`` but a ``void *token`` argument,
+and the same behaviour except matching tokens, rather than only defs.
+
+
+New slots
+---------
+
+For each field of the ``PyModuleDef`` struct, except ones from
+``PyModuleDef_HEAD_INIT``, a new slot ID will be provided: ``Py_mod_name``,
+``Py_mod_doc``, ``Py_mod_clear``, etc.
+Slots related to the module state rather than the module object will
+use a ``Py_mod_state_`` prefix.
+See :ref:`pep793-api-summary` for a full list.
+
+All new slots -- these and ``Py_tp_token`` discussed above -- may not be
+repeated in the slots array, and may not be used in a
+``PyModuleDef.m_slots`` array.
+They may not have a ``NULL`` value (instead, the slot can be omitted entirely).
+
+Note that currently, for modules created from a *spec* (that is, using
+``PyModule_FromDefAndSpec``), the ``PyModuleDef.m_name`` member is ignored
+and the name from the spec is used instead.
+All API proposed in this document creates modules from a *spec*, and it will
+ignore ``Py_mod_name`` in the same way.
+The slot will be optional, but extensions will be strongly encouraged to
+include it for the benefit of future APIs, external tooling,
+and debugging/introspection.
+
+
+Bits & Pieces
+-------------
+
+A ``PyMODEXPORT_FUNC`` macro will be added, similar to the ``PyMODINIT_FUNC``
+macro but with ``PyModuleDef_Slot *`` as the return type.
+
+A ``PyModule_GetStateSize`` function will be added to retrieve the size set
+by ``Py_mod_state_size`` or ``PyModuleDef.m_size``.
+Since the result may be -1 (for single-phase-init modules), it will be output
+via a pointer; the function will return 0 on success and -1 on failure:
+
+.. code-block:: c
+
+    int PyModule_GetStateSize(PyObject *, Py_ssize_t *result);
+
+
+.. _pep793-api-summary:
+
+New API summary
+---------------
+
+The following functions will be added:
+
+.. code-block:: c
+
+    PyObject *PyModule_FromSlotsAndSpec(PyModuleDef_Slot *, PyObject *spec)
+    int PyModule_Exec(PyObject *)
+    int PyModule_GetToken(PyObject *, void**)
+    PyObject *PyType_GetModuleByToken(PyTypeObject *type, void *token)
+    int PyModule_GetStateSize(PyObject *, Py_ssize_t *result);
+
+A new macro will be added:
+
+.. code-block:: c
+
+    PyMODEXPORT_FUNC
+
+And new slot types (``#define``\d names for small integers):
+
+- ``Py_mod_name`` (equivalent to ``PyModuleDef.m_name``)
+- ``Py_mod_doc`` (equivalent to ``PyModuleDef.m_doc``)
+- ``Py_mod_state_size`` (equivalent to ``PyModuleDef.m_size``)
+- ``Py_mod_methods`` (equivalent to ``PyModuleDef.m_methods``)
+- ``Py_mod_state_traverse`` (equivalent to ``PyModuleDef.m_traverse``)
+- ``Py_mod_state_clear`` (equivalent to ``PyModuleDef.m_clear``)
+- ``Py_mod_state_free`` (equivalent to ``PyModuleDef.m_free``)
+- ``Py_mod_token`` (see above)
+
+All this will be added to the Limited API.
+
+
+Backwards Compatibility
+=======================
```

Comment:
I think there should be greater emphasis on compatibility code for extension modules supporting older versions, but wanting to use the better/newer mechanism for 3.15+. Either in *Backwards Compatibility* or *How to Teach This*.

For instance, the multiple `Py_mod_exec` is solvable with a new function that calls all existing functions sequentially. Any suggested mechanisms `#if ...` for defining both a `PyInit` and `PyModExport` export hook? Anything that could go into pythoncapi-compat to make transition easier?